sycamore/functions/rabin_karp.py

```diff
@@ -0,0 +1,88 @@
+from functools import reduce
+
+
+class RkHash:
+    """
+    Class RkHash encapsulates a 64-bit Rabin-Karp hash function.  It is
+    meant to hash one byte [0,255] at a time.  It's not the best hash
+    function, but has the nice property of allowing previous inputs to
+    be removed.  This makes it efficient for sliding-window applications.
+    """
+
+    __slots__ = ["shift", "prime", "primeShift", "val", "inv"]
+
+    def __init__(self, width: int) -> None:
+        if width < 1:
+            raise ValueError
+        self.shift = 8
+        self.prime = 36028797018963913  # largest prime < 2^55
+        self.primeShift = self.prime * self.shift
+        self.val = 0
+        self.inv = reduce(lambda x, _: (x << self.shift) % self.prime, range(width - 1), 1)
+
+    def __str__(self) -> str:
+        return "(rk:%d:%d)" % (self.val, self.inv)
```

Comment:
You are of course free to format strings however you want, but just in case you hadn't seen them, I find Pythons [f-strings](https://docs.python.org/3/reference/lexical_analysis.html#f-strings) to be one of the biggest quality of life improvements in recent versions. Course that may just be that I also forget the stupid % codes.

Reply:
sprintf-style (%) formatting of ints is faster than format(), which is what f-strings are.  One nice thing about f-strings, though, is that they support the 'b' type for binary output, which % doesn't for some reason.  https://stackoverflow.com/questions/40714555/python-string-formatting-is-more-efficient-than-format-function